### DISCUSSION TITLE
Pruned FFTs

### DISCUSSION THREAD

**Discussion by jaemolihm** [wannier-berri/wannier-berri#215] [General]
I learned that FFT with only few nonzero elements in the input (or vice versa) - which I was personally calling the "Wannier Berri trick" - are called "pruned FFTs"! (Unfortunate that prune is not a berri 😂)

https://www.fftw.org/pruned.html

> Computing only the first K outputs of a DFT with FFTW
> An important special case where one can easily achieve O(N log K) complexity is where one wishes to compute only the first K outputs of a DFT, where K divides N.

> In this case, one simply computes N/K FFTs of size K, with stride N/K in the data, and then combines the output by summing with appropriate phase ("twiddle") factors. Essentially, you are performing "by hand" a single radix-K decimation-in-frequency Cooley-Tukey step, where you only compute one size-K block of the output.

> (The dual problem is where only the first K inputs are non-zero, in which case you essentially just reverse these steps.)

**Comment by stepan-tsirkin** [comment]
Great to know that I reinvented a wheel (which I always suspected, btw)

Looks like the oldest reference is dated 1971 :  [J. Markel, "FFT pruning," in IEEE Transactions on Audio and Electroacoustics, vol. 19, no. 4, pp. 305-311, December 1971, doi: 10.1109/TAU.1971.1162205.](https://ieeexplore.ieee.org/document/1162205)

If I knew it before, I would cite some of those papers, but know I will add some links to the web page.
